### ISSUE TITLE
Quantization-aware training with non-parameter layers and BN layers?

### ISSUE THREAD

**Issue by Wanwannodao** [NervanaSystems/distiller#256]
According to [this paper](https://arxiv.org/abs/1806.08342), we should accurately simulate quantization related effects such as fusion of operations and batch normalization folding during quant-aware training.
`PostTrainLInearQuantizer` seems to replace layers with no parameter (e.g. `torch.cat`)  by the quantized version of them (e.g. `RangeLinearQuantConcatWrapper`).

 https://github.com/NervanaSystems/distiller/blob/fac2359bbd075879aa633aa6cfd57981bee1e250/distiller/quantization/range_linear.py#L797-L802

However, in `QuantAwareTrainRangeLienarQuantizer`, there is no such handling for non-parameter layers.

So, my question is that, when I use layers like concat, eltwise-* in quant-aware training, 
1. should I add such replacements ? and 
2. how can I simulate the batch normalization folding behavior?

Thank you !

**Comment by guyjacob** [comment]
Hi @Wanwannodao 

Indeed, as it is `QuantAwareTrainRangeLinearQuantizer` only fake-quantizes the activations of `nn.ReLU` modules. This essentially means that if we have the following sequence:

relu_0 --> layer_x --> layer_x+1 --> ... --> layer_x+n --> relu_1

Then, in the context of activations quantization, the layers [layer_x, layer_x+1, ..., layer_x+n] are assumed to be fused with "relu_1".
You'd be correct in saying this covers only a specific (albeit a very common) pattern.

If you have cases where you know you need to insert fake-quantization nodes, then at the moment I can offer the following:
1. Sub-class `QuantAwareTrainRangeLinearQuantizer`
2. In your sub-class, add replacement factory entries for the layer type whose output you wish to quantize.
3. If you need to quantize only specific instances of that layer type, then take care of that using the overrides mechanism.

(You could of course just modify `QuantAwareTrainRangeLinearQuantizer` directly instead of sub-classing)
Let me know if you need more help with this.

Regarding BN folding - this has been raised before, see #166. Unfortunately we haven't gotten to implementing this yet. I hope we can get to it soon, but can't say at the moment when that would be. If this is something you think you'd like to tackle, let us know.

**Comment by Wanwannodao** [comment]
Hi, thank you for your answering !

That means, for example, the following pattern isn't supported for now, is it ?

relu_0 -- \\
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp;&nbsp; add_x --> relu_2 ... 
relu_1 -- /  

I mean, there are layer types whose input tensors have to be requantized to match the scaling factors and the zero-points, and output tensor may not require quantization, right ?  
In such a case, what should I do ? Defining another fake-quantization wrapper with `forward` looking like below ?

```python
def forward(self, *input):
    res = self.fake_requant(res)
    res = self.wrapped_module(*input)
    return res
```



And, yes, I will probably need to implement BN folding.

**Comment by guyjacob** [comment]
Regarding the pattern you drew - the code you wrote is a start, indeed you'd need to have fake quant operation before the element-wise add module. But, in the case of element wise addition, the inputs are re-quantized according to the scale of the **output**. So the fake requant operation will need to get some info from the fake requant op in the output.
All of that requires a way to understand the order of execution of the model. We have the infrastructure for that with the SummaryGraph class, it's just not readily available for the quantizer. This is WIP, I hope to push it sometime next week.

Regarding BN folding - we will take a look at it from our side. If we end up deciding we want help, I'll update here.

**Comment by Wanwannodao** [comment]
Hi.
Ah, that's true... It requires more steps... Thanks for your explanation.

Okay. No problem.

**Comment by levzlotnik** [comment]
Hi,
BatchNorm folding has been implemented and merged.
This issue seems resolved, and hence I'm closing it. Feel free to reopen if needed.

**Comment by vdgs** [comment]
In this case: conv->conv->conv->relu, we need convolution output scale/zero_point to convert output to  int8/uint8 for accelerating speed, how to solve it ?
